src/rtc_stream.cpp

```diff
@@ -0,0 +1,274 @@
+	rhs.m_data_channel.reset();
+
+	std::swap(m_incoming, rhs.m_incoming);
+	std::swap(m_incoming_size, rhs.m_incoming_size);
+}
+
+rtc_stream::~rtc_stream()
+{
+	close();
+}
+
+void rtc_stream::on_message(error_code const& ec, std::vector<char> data)
+{
+	if(ec)
+	{
+		// Ignore
+		return;
+	}
+
+	m_incoming_size += data.size();
+	m_incoming.emplace(std::move(data));
+
+	// Fulfil pending read if any
+	if(m_read_handler) issue_read();
+}
+
+close_reason_t rtc_stream::get_close_reason()
+{
+	return close_reason_t::none;
+}
+
+void rtc_stream::close()
+{
+	if(m_data_channel) // it can be null after a move constructor
+	{
+		m_data_channel->onMessage([](std::variant<rtc::binary, rtc::string> const&) {});
+		m_data_channel->close();
+	}
+
+	cancel_handlers(boost::asio::error::operation_aborted);
+}
+
+void rtc_stream::cancel_handlers(error_code const& ec)
+{
+	TORRENT_ASSERT(ec);
+
+	if(m_read_handler) m_read_handler(ec, 0);
+	if(m_write_handler) m_write_handler(ec, 0);
+
+	m_read_handler = nullptr;
+	m_read_buffer.clear();
+	m_read_buffer_size = 0;
+
+	m_write_handler = nullptr;
+	m_write_buffer.clear();
+	m_write_buffer_size = 0;
+}
+
+bool rtc_stream::ensure_open()
+{
+	if(is_open()) return true;
+
+    cancel_handlers(boost::asio::error::not_connected);
+    return false;
+}
+
+bool rtc_stream::is_open() const
+{
+	return !m_data_channel->isClosed();
+}
+
+std::size_t rtc_stream::available() const
+{
+	return m_incoming_size;
+}
+
+rtc_stream::endpoint_type rtc_stream::remote_endpoint(error_code& ec) const
+{
+    if (!is_open())
+    {
+        ec = boost::asio::error::not_connected;
+        return endpoint_type();
+    }
+
+	auto addr = m_peer_connection->remoteAddress();
+	if(!addr)
+	{
+		ec = boost::asio::error::operation_not_supported;
+		return endpoint_type();
+	}
+
+	size_t pos = addr->find_last_of(':');
+	if(pos == std::string::npos)
+	{
+		ec = boost::asio::error::address_family_not_supported;
+		return endpoint_type();
+	}
+
+	return endpoint_type(ip::make_address(addr->substr(0, pos), ec)
+			, std::stoul(addr->substr(pos+1)));
+}
+
+rtc_stream::endpoint_type rtc_stream::local_endpoint(error_code& ec) const
+{
+	if (!is_open())
+    {
+        ec = boost::asio::error::not_connected;
+        return endpoint_type();
+    }
+
+	auto addr = m_peer_connection->localAddress();
+	if(!addr)
+	{
+		ec = boost::asio::error::operation_not_supported;
+		return endpoint_type();
+	}
+
+	size_t pos = addr->find_last_of(':');
+	if(pos == std::string::npos)
+	{
+		ec = boost::asio::error::address_family_not_supported;
+		return endpoint_type();
+	}
+
+	return endpoint_type(ip::make_address(addr->substr(0, pos), ec)
+			, std::stoul(addr->substr(pos+1)));
+}
+
+void rtc_stream::issue_read()
+{
+	TORRENT_ASSERT(m_read_handler);
+	TORRENT_ASSERT(m_read_buffer_size > 0);
+
+	if(!ensure_open()) return;
+
+	std::size_t bytes_read = read_some();
+	if(bytes_read > 0)
+	{
+		post(m_io_context, std::bind(m_read_handler, error_code{}, bytes_read));
```

Comment:
this is a bit of a nit-pick: imagine `post()` throwing an exception, then you won't clear `m_read_handler` or `m_read_buffer`. I would suggest (in general) clearing the state before posting the callback.
In order to clear the handler and post it you can use `std::exchange()`. Like this:
```
post(m_io_context, std::bind(std::exchange(m_read_handler, nullptr), error_code{}, bytes_read));
```

That way, you clear the handler before calling `post()`